price-estimator/src/metrics.rs

```diff
@@ -0,0 +1,64 @@
+use anyhow::Result;
+use prometheus::{Histogram, HistogramOpts, HistogramVec, IntCounterVec, Opts, Registry};
+use std::time::Instant;
+use warp::log::Info;
+
+// There are global metrics (response_status and response_time) that are measured through warp's log
+// functionality. The global metrics work on the final response and do not know about the specific
+// route that handled them.
```

Comment:
The Info struct contains the route as a field. Could we use that to log the route from the top level filter instead of hooking into each handler individually?

Reply:
We definitely want to label per route because the response times are so different and for the same reason also separate success responses from errors.

Using the path from info was my first attempt but it has problems: We would need to parse the url again to figure out which route is being used. There is no nice way to do this except by manually inspecting url. I couldn't find a way to reuse the filters for this. So we end up writing the path handling code again, mirroring our structure from the filters.

Could you explain more why you prefer using only the top level? I don't see how double counting the response times is a problem. The response times are recorded in separate, partially overlapping metrics but that doesn't make the individual metric wrong.

Reply:
I also would like the metrics handling to be at a top level (not necessarily in `main.rs`, but maybe in `filters::all`?), just to reduce the amount of plumbing required. I also really like the labeled response metrics since, as you said, these requests have vastly different response times. I would even go as far as suggesting that other things (such as query parameters) should be labeled as well (for example, there is a significant overhead to querying with `blockNumber` or `batchId` parameters). 

Maybe the sub-filters like `markets` can return a `impl Filter<Extract = (impl Reply, ResponseLabel)>` that so the metrics handling only needs to done in one place, and only the labels need to be returned by the filters? This way we don't need to actually modify the handler code to add metrics, only the filters around them - which I find is a nice and clean solution. I imagine something like `struct ResponseLabel { name: &'static str, time: EstimationTime }` or something, basically including all the important things that need to be labeled for recording response timing.

Reply:
> for example, there is a significant overhead to querying with blockNumber or batchId parameters

Yes I thought about this too. We probably want to label "cached" or "now" versus "uncached" or "custom" or something like that. Actually I'm not even sure if the prod price estimator should expose those routes because it's so easy to dos with them and real users do not need them. But that's a bigger discussion.

I am going to try your idea with returning more information from the sub filters.

Reply:
> Could you explain more why you prefer using only the top level?

Mainly for convenience reasons (having to remember to add logging to each new filter) and reasoning about which metric is which (requires understanding why there are two and what the difference is).

Do you have a concern with just logging the entire path as a label? We can aggregate the ones we want using regexs in Graphana later. 
I remember we shouldn't have a completely unbound number of labels for prometheus (since each label is another series), but it feels like for us it should "only" be O(n^2) the number of tokens (due to how markets scale).

Reply:
That's already too many labels and it's not taking into account that there are multiple ways to refer to the same token. Also it would include random bogus 404 urls that someone might throw at the service, and we would still have to parse the url to remove the query.

I think Nick's idea works though (still trying to make it a bit nicer).